File: WindFlow/YSB/src/ysb.cpp
/**************************************************************************************
 *  Copyright (c) 2019- Gabriele Mencagli
 *  
 *  This file is part of StreamBenchmarks.
 *  
 *  StreamBenchmarks is free software dual licensed under the GNU LGPL or MIT License.
 *  You can redistribute it and/or modify it under the terms of the
 *    * GNU Lesser General Public License as published by
 *      the Free Software Foundation, either version 3 of the License, or
 *      (at your option) any later version
 *    OR
 *    * MIT License: https://github.com/ParaGroup/StreamBenchmarks/blob/master/LICENSE.MIT
 *  
 *  StreamBenchmarks is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *  You should have received a copy of the GNU Lesser General Public License and
 *  the MIT License along with WindFlow. If not, see <http://www.gnu.org/licenses/>
 *  and <http://opensource.org/licenses/MIT/>.
 **************************************************************************************
 */

#include <regex>
#include <string>
#include <vector>
#include <iostream>
#include <ff/ff.hpp>
#include <windflow.hpp>

#include "../includes/nodes/sink.hpp"
#include "../includes/util/event.hpp"
#include "../includes/util/result.hpp"
#include "../includes/nodes/source.hpp"
#include "../includes/nodes/filter.hpp"
#include "../includes/nodes/joiner.hpp"
#include "../includes/util/cli_util.hpp"
#include "../includes/util/constants.hpp"
#include "../includes/util/joined_event.hpp"
#include "../includes/util/campaign_generator.hpp"

using namespace std;
using namespace ff;
using namespace wf;
using namespace chrono;

// global variables
atomic<long> sent_tuples;                    // total number of events processed by the system

// function for computing the final aggregates on tumbling windows (INCremental version)
void aggregateFunctionINC(const joined_event_t &event, result_t &result, RuntimeContext &rc)
{
    result.count++;
    if (event.ts > result.lastUpdate) {
        result.lastUpdate = event.ts;
    }
}

// Main
int main(int argc, char* argv[]) {
    /// parse arguments from command line
    int option = 0;
    int index = 0;
    size_t source_par_deg = 0;
    size_t filter_par_deg = 0;
    size_t joiner_par_deg = 0;
    size_t winAgg_par_deg = 0;
    size_t sink_par_deg = 0;
    int rate = 0;
    sent_tuples = 0;
    long sampling = 0;
    bool chaining = false;
    size_t batch_size = 0;
    if (argc == 9 || argc == 10) {
        while ((option = getopt_long(argc, argv, "r:s:p:b:c:", long_opts, &index)) != -1) {
            switch (option) {
                case 'r': {
                    rate = atoi(optarg);
                    break;
                }
                case 's': {
                    sampling = atoi(optarg);
                    break;
                }
                case 'b': {
                    batch_size = atoi(optarg);
                    break;
                }
                case 'p': {
                    vector<size_t> par_degs;
                    string pars(optarg);
                    stringstream ss(pars);
                    for (size_t i; ss >> i;) {
                        par_degs.push_back(i);
                        if (ss.peek() == ',')
                            ss.ignore();
                    }
                    if (par_degs.size() != 5) {
                        printf("Error in parsing the input arguments\n");
                        exit(EXIT_FAILURE);
                    }
                    else {
                        source_par_deg = par_degs[0];
                        filter_par_deg = par_degs[1];
                        joiner_par_deg = par_degs[2];
                        winAgg_par_deg = par_degs[3];
                        sink_par_deg = par_degs[4];
                    }
                    break;
                }
                case 'c': {
                    chaining = true;
                    break;
                }
                default: {
                    printf("Error in parsing the input arguments\n");
                    exit(EXIT_FAILURE);
                }
            }
        }
    }
    else if (argc == 2) {
        while ((option = getopt_long(argc, argv, "h", long_opts, &index)) != -1) {
            switch (option) {
                case 'h': {
                    printf("Parameters: --rate <value> --sampling <value> --batch <size> --parallelism <nSource,nFilter,nJoiner,nWinAggregate,nSink> [--chaining]\n");
                    exit(EXIT_SUCCESS);
                }
            }
        }
    }
    else {
        printf("Error in parsing the input arguments\n");
        exit(EXIT_FAILURE);
    }
    // create the campaigns
    CampaignGenerator campaign_gen;
    /// application starting time
    unsigned long app_start_time = current_time_nsecs();
    cout << "Executing YSB with parameters:" << endl;
    if (rate != 0) {
        cout << "  * rate: " << rate << " tuples/second" << endl;
    }
    else {
        cout << "  * rate: full_speed tupes/second" << endl;
    }
    cout << "  * batch size: " << batch_size << endl;
    cout << "  * sampling: " << sampling << endl;
    cout << "  * source: " << source_par_deg << endl;
    cout << "  * filter: " << filter_par_deg << endl;
    cout << "  * joiner: " << joiner_par_deg << endl;
    cout << "  * winAggregate: " << winAgg_par_deg << endl;
    cout << "  * sink: " << sink_par_deg << endl;
    cout << "  * topology: source -> filter -> joiner -> winAggregate -> sink" << endl;
    PipeGraph topology(topology_name, Execution_Mode_t::DEFAULT, Time_Policy_t::INGRESS_TIME);
    if (!chaining) { // no chaining
        /// create the operators
        Source_Functor source_functor(rate, app_start_time, campaign_gen.getArrays(), campaign_gen.getAdsCompaign());
        Source source = Source_Builder(source_functor)
                .withParallelism(source_par_deg)
                .withName(source_name)
                .withOutputBatchSize(batch_size)
                .build();
        Filter_Functor filter_functor;
        Filter filter = Filter_Builder(filter_functor)
                .withParallelism(filter_par_deg)
                .withName(filter_name)
                .withOutputBatchSize(batch_size)
                .build();
        Joiner_Functor joiner_functor(campaign_gen.getHashMap(), campaign_gen.getRelationalTable());
        FlatMap joiner = FlatMap_Builder(joiner_functor)
                .withParallelism(joiner_par_deg)
                .withName(joiner_name)
                .withOutputBatchSize(batch_size)
                .build();
        Keyed_Windows winAggregate = Keyed_Windows_Builder(aggregateFunctionINC)
                .withTBWindows(seconds(10), seconds(10))
                .withName("yb_kf")
                .withParallelism(winAgg_par_deg)
                .withKeyBy([](const joined_event_t &in) -> unsigned long { return in.cmp_id; })
                .withOutputBatchSize(batch_size)
                .build();
        Sink_Functor sink_functor(sampling, app_start_time);
        Sink sink = Sink_Builder(sink_functor)
                .withParallelism(sink_par_deg)
                .withName(sink_name)
                .build();
        MultiPipe &mp = topology.add_source(source);
        cout << "Chaining is disabled" << endl;
        mp.add(filter);
        mp.add(joiner);
        mp.add(winAggregate);
        mp.add_sink(sink);
    }
    else { // chaining
        /// create the operators
        Source_Functor source_functor(rate, app_start_time, campaign_gen.getArrays(), campaign_gen.getAdsCompaign());
        Source source = Source_Builder(source_functor)
                .withParallelism(source_par_deg)
                .withName(source_name)
                .build();
        Filter_Functor filter_functor;
        Filter filter = Filter_Builder(filter_functor)
                .withParallelism(filter_par_deg)
                .withName(filter_name)
                .build();
        Joiner_Functor joiner_functor(campaign_gen.getHashMap(), campaign_gen.getRelationalTable());
        FlatMap joiner = FlatMap_Builder(joiner_functor)
                .withParallelism(joiner_par_deg)
                .withName(joiner_name)
                .withOutputBatchSize(batch_size)
                .build();
        Keyed_Windows winAggregate = Keyed_Windows_Builder(aggregateFunctionINC)
                .withTBWindows(seconds(10), seconds(10))
                .withName("yb_kf")
                .withParallelism(winAgg_par_deg)
                .withKeyBy([](const joined_event_t &in) -> unsigned long { return in.cmp_id; })
                .build();
        Sink_Functor sink_functor(sampling, app_start_time);
        Sink sink = Sink_Builder(sink_functor)
                .withParallelism(sink_par_deg)
                .withName(sink_name)
                .build();
        MultiPipe &mp = topology.add_source(source);
        cout << "Chaining is enabled" << endl;
        mp.chain(filter);
        mp.chain(joiner);
        mp.add(winAggregate);
        mp.chain_sink(sink);
    }
    cout << "Executing topology" << endl;
    /// evaluate topology execution time
    volatile unsigned long start_time_main_usecs = current_time_usecs();
    topology.run();
    volatile unsigned long end_time_main_usecs = current_time_usecs();
    cout << "Exiting" << endl;
    double elapsed_time_seconds = (end_time_main_usecs - start_time_main_usecs) / (1000000.0);
    double throughput = sent_tuples / elapsed_time_seconds;
    cout << "Measured throughput: " << (int) throughput << " tuples/second" << endl;
    cout << "Dumping metrics" << endl;
    util::metric_group.dump_all();
    return 0;
}

File: Flink/YSB/src/main/java/YSB/YSB.java
/**************************************************************************************
 *  Copyright (c) 2019- Gabriele Mencagli
 *  
 *  This file is part of StreamBenchmarks.
 *  
 *  StreamBenchmarks is free software dual licensed under the GNU LGPL or MIT License.
 *  You can redistribute it and/or modify it under the terms of the
 *    * GNU Lesser General Public License as published by
 *      the Free Software Foundation, either version 3 of the License, or
 *      (at your option) any later version
 *    OR
 *    * MIT License: https://github.com/ParaGroup/StreamBenchmarks/blob/master/LICENSE.MIT
 *  
 *  StreamBenchmarks is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *  You should have received a copy of the GNU Lesser General Public License and
 *  the MIT License along with WindFlow. If not, see <http://www.gnu.org/licenses/>
 *  and <http://opensource.org/licenses/MIT/>.
 **************************************************************************************
 */

package YSB;

import Util.Log;
import java.util.*;
import Util.MetricGroup;
import org.slf4j.Logger;
import java.io.IOException;
import Constants.YSBConstants;
import Util.ThroughputCounter;
import org.apache.flink.util.*;
import org.slf4j.LoggerFactory;
import Constants.BaseConstants;
import Constants.YSBConstants.Conf;
import java.util.concurrent.TimeUnit;
import Constants.YSBConstants.Component;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.api.common.JobExecutionResult;
import org.apache.flink.api.java.typeutils.TypeExtractor;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.api.common.functions.FoldFunction;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

// class YSB
public class YSB {
    private static final Logger LOG = Log.get(YSB.class);

    // main
    public static void main(String[] args) {
        if (args.length == 1 && args[0].equals(BaseConstants.HELP)) {
            String alert = "Parameters: --rate <value> --sampling <value> --parallelism <nSource nFilter nJoiner nAggregate nSink> [--chaining]\n";
            LOG.error(alert);
        }
        else {
            // load configuration
            ParameterTool params;
            Configuration conf;
            try {
                params = ParameterTool.fromPropertiesFile(YSB.class.getResourceAsStream(YSBConstants.DEFAULT_PROPERTIES));
                conf = params.getConfiguration();
                LOG.debug("Loaded configuration file");
            }
            catch (IOException e) {
                LOG.error("Unable to load configuration file", e);
                throw new RuntimeException("Unable to load configuration file", e);
            }
            // parse command line arguments
            boolean isCorrect = true;
            int gen_rate = -1;
            int sampling = 1;
            int source_par_deg = 1;
            int filter_par_deg = 1;
            int joiner_par_deg = 1;
            int agg_par_deg = 1;
            int sink_par_deg = 1;
            boolean isChaining = false;
            if (args.length == 10 || args.length == 11) {
                if (!args[0].equals("--rate")) {
                    isCorrect = false;
                }
                else {
                    try {
                        gen_rate = Integer.parseInt(args[1]);
                    }
                    catch (NumberFormatException e) {
                        isCorrect = false;
                    }
                }
                if (!args[2].equals("--sampling"))
                    isCorrect = false;
                else {
                    try {
                        sampling = Integer.parseInt(args[3]);
                    }
                    catch (NumberFormatException e) {
                        isCorrect = false;
                    }
                }
                if (!args[4].equals("--parallelism"))
                    isCorrect = false;
                else {
                    try {
                        source_par_deg = Integer.parseInt(args[5]);
                        filter_par_deg = Integer.parseInt(args[6]);
                        joiner_par_deg = Integer.parseInt(args[7]);
                        agg_par_deg = Integer.parseInt(args[8]);
                        sink_par_deg = Integer.parseInt(args[9]);
                    }
                    catch (NumberFormatException e) {
                        isCorrect = false;
                    }
                }
                if (args.length == 11) {
                    if (!args[10].equals("--chaining")) {
                        isCorrect = false;
                    }
                    else {
                        isChaining = true;
                    }
                }
            }
            else {
                LOG.error("Error in parsing the input arguments");
                System.exit(1);
            }
            if (!isCorrect) {
               LOG.error("Error in parsing the input arguments");
               System.exit(1);
            }
            String topology_name = YSBConstants.DEFAULT_TOPO_NAME;
            long runTimeSec = conf.getInteger(Conf.RUNTIME, 0);
            // initialize the HashMap of campaigns
            int numCampaigns = conf.getInteger(Conf.NUM_KEYS, 0);
            List<CampaignAd> campaignAdSeq = generateCampaignMapping(numCampaigns);
            HashMap<String, String> campaignLookup = new HashMap<String, String>();
            for (int i=0; i<campaignAdSeq.size(); i++) {
                campaignLookup.put((campaignAdSeq.get(i)).ad_id, (campaignAdSeq.get(i)).campaign_id);
            }

            // create the execution environment
            StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
            env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

            // flush as soon as possible in throttled mode (minimize the latency)
            if (gen_rate != 0) {
                env.setBufferTimeout(0);
            }

            // tag for late tuples
            final OutputTag<Joined_Event> lateOutputTag = new OutputTag<Joined_Event>("late-data"){};

            // create the topology
            DataStream<YSB_Event> event_stream = env.addSource(new YSBSource(gen_rate, source_par_deg, runTimeSec, campaignAdSeq, numCampaigns)).setParallelism(source_par_deg);

            DataStream<YSB_Event> filtered_stream = event_stream.filter(new Filter(filter_par_deg)).setParallelism(filter_par_deg);

            DataStream<Joined_Event> joined_stream = filtered_stream.flatMap(new Joiner(campaignLookup, joiner_par_deg)).setParallelism(joiner_par_deg);

            SingleOutputStreamOperator<Aggregate_Event> agg_stream = joined_stream.keyBy("cmp_id").window(TumblingEventTimeWindows.of(Time.seconds(10))).sideOutputLateData(lateOutputTag).fold(new Aggregate_Event(), new FoldFunction<Joined_Event, Aggregate_Event> () {
                public Aggregate_Event fold(Aggregate_Event agg, Joined_Event event) {
                    agg.cmp_id = event.cmp_id;
                    agg.count++;
                    agg.timestamp = event.timestamp;
                    return agg;
                }
            }).setParallelism(agg_par_deg);

            DataStream<Joined_Event> lateStream = agg_stream.getSideOutput(lateOutputTag);

            lateStream.addSink(new DroppedSink()).setParallelism(1);

            agg_stream.addSink(new ConsoleSink(sink_par_deg, gen_rate, sampling)).setParallelism(sink_par_deg);

            // print app info
            LOG.info("Executing YSB with parameters:\n" +
                     "  * rate: " + ((gen_rate == 0) ? "full_speed" : gen_rate) + " tuples/second\n" +
                     "  * sampling: " + sampling + "\n" +
                     "  * source: " + source_par_deg + "\n" +
                     "  * filter: " + filter_par_deg + "\n" +
                     "  * joiner: " + joiner_par_deg + "\n" +
                     "  * winAggregate: " + agg_par_deg + "\n" +
                     "  * sink: " + sink_par_deg + "\n" +
                     "  * topology: source -> filter -> joiner -> winAggregate -> sink");

            // run the topology
            try {
                // configure the environment
                if (!isChaining) {
                    env.disableOperatorChaining();
                    LOG.info("Chaining is disabled");
                }
                else {
                    LOG.info("Chaining is enabled");
                }
                // run the topology
                LOG.info("Submitting topology");
                JobExecutionResult result = env.execute();
                LOG.info("Exiting");
                // measure throughput
                double throughput = (double) (ThroughputCounter.getValue() / result.getNetRuntime(TimeUnit.SECONDS));
                LOG.info("Measured throughput: " + (int) throughput + " tuples/second");
                // dump the metrics
                LOG.info("Dumping metrics");
                MetricGroup.dumpAll();
            }
            catch (Exception e) {
                LOG.error(e.toString());
            }
        }
    }

    // generate in-memory ad_id to campaign_id map
    private static List<CampaignAd> generateCampaignMapping(int numCampaigns) {
        CampaignAd[] campaignArray = new CampaignAd[numCampaigns*10];
        for (int i=0; i<numCampaigns; i++) {
            String campaign = UUID.randomUUID().toString();
            for (int j=0; j<10; j++) {
                campaignArray[(10*i)+j] = new CampaignAd(UUID.randomUUID().toString(), campaign);
            }
        }
        return Arrays.asList(campaignArray);
    }
}
